File: core/src/de/project/ice/IceGame.java
package de.project.ice;

import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.I18NBundle;
import de.project.ice.dialog.Dialog;
import de.project.ice.ecs.IceEngine;
import de.project.ice.hotspot.HotspotManager;
import de.project.ice.inventory.Inventory;
import de.project.ice.screens.*;
import de.project.ice.utils.Assets;
import de.project.ice.utils.Pair;
import de.project.ice.utils.SceneLoader;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class IceGame extends ApplicationAdapter
{
    private final Array<BaseScreen> screens = new Array<BaseScreen>();
    private final Array<BaseScreen> screensToAdd = new Array<BaseScreen>();
    private final Array<Pair<BaseScreen, Boolean>> screensToRemove = new Array<Pair<BaseScreen, Boolean>>();
    public IceEngine engine;
    protected GameScreen gameScreen = null;
    @NotNull
    protected CursorScreen cursorScreen;
    @NotNull
    public Inventory inventory;
    @NotNull
    public HotspotManager hotspotManager;
    @NotNull
    public I18NBundle strings;
    private boolean paused = false;

    public IceGame()
    {
    }

    @Override
    public void create()
    {
        I18NBundle.setSimpleFormatter(true);
        inventory = new Inventory(this);
        hotspotManager = new HotspotManager(this);
        engine = new IceEngine(this);
        strings = I18NBundle.createBundle(Gdx.files.internal("strings/ProjectIce"));

        gameScreen = new GameScreen(this, engine);
        addScreen(gameScreen);
        cursorScreen = new CursorScreen(this);
        addScreen(cursorScreen);
        init();

    }

    protected void init()
    {
        addScreen(new MainMenuScreen(this));
        Gdx.input.setInputProcessor(new InputMultiplexer());
    }

    public void pauseGame()
    {
        if (gameScreen != null)
        {
            gameScreen.pauseGame();
            paused = true;
        }
    }

    public void resumeGame()
    {
        if (gameScreen != null)
        {
            gameScreen.resumeGame();
            paused = false;
        }
    }

    public boolean isGamePaused()
    {
        return paused;
    }

    public void showDialog(String dialog)
    {
        addScreen(new DialogScreen(this, Dialog.load(Gdx.files.internal("dialog/" + dialog + ".dlz"))));
    }

    public void showMessages(String... messages)
    {
        addScreen(new MessageScreen(this, messages));
    }

    @Override
    public void dispose()
    {
        for (BaseScreen screen : screens)
        {
            screen.hide();
            screen.dispose();
        }
        screens.clear();
    }

    @Override
    public void pause()
    {
        if (screens.size > 0)
        {
            screens.peek().pause();
        }
    }

    @Override
    public void resume()
    {
        if (screens.size > 0)
        {
            screens.peek().resume();
        }
    }

    @NotNull
    public CursorScreen.Cursor getPrimaryCursor()
    {
        return cursorScreen.getPrimaryCursor();
    }

    public void setPrimaryCursor(@NotNull CursorScreen.Cursor primaryCursor)
    {
        cursorScreen.setPrimaryCursor(primaryCursor);
    }

    @NotNull
    public CursorScreen.Cursor getSecondaryCursor()
    {
        return cursorScreen.getSecondaryCursor();
    }

    public void setSecondaryCursor(@NotNull CursorScreen.Cursor secondaryCursor)
    {
        cursorScreen.setSecondaryCursor(secondaryCursor);
    }

    @Override
    public void render()
    {
        Assets.update();

        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT | GL20.GL_DEPTH_BUFFER_BIT);
        float delta = Gdx.graphics.getDeltaTime();
        for (BaseScreen screen : screens)
        {
            screen.update(delta);
        }
        for (BaseScreen screen : screens)
        {
            screen.render();
        }
        for (BaseScreen screen : screensToAdd)
        {
            int index = -1;
            for (int i = 0; i < screens.size; i++)
            {
                if (screens.get(i).getPriority() < screen.getPriority())
                {
                    index = i;
                    break;
                }
            }
            if (index == -1)
            {
                if (screens.size > 0)
                {
                    screens.peek().pause();
                }
                screens.add(screen);
                screen.show();
            }
            else
            {
                screens.insert(index, screen);
            }
            screen.resize(Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
        }
        screensToAdd.clear();
        for (Pair<BaseScreen, Boolean> pair : screensToRemove)
        {
            if (pair == screens.peek())
            {
                screens.pop();
                screens.peek().resume();
            }
            else
            {
                screens.removeValue(pair.getFirst(), true);
            }
            pair.getFirst().hide();
            if (pair.getSecond())
            {
                pair.getFirst().dispose();
            }
        }
        screensToRemove.clear();
    }

    @Override
    public void resize(int width, int height)
    {
        for (BaseScreen screen : screens)
        {
            screen.resize(width, height);
        }
    }

    public void removeScreen(@NotNull BaseScreen screen)
    {
        removeScreen(screen, true);
    }

    public void removeScreen(@NotNull BaseScreen screen, boolean dispose)
    {
        screensToRemove.add(new Pair<BaseScreen, Boolean>(screen, dispose));
    }

    public void addScreen(@NotNull BaseScreen screen)
    {
        screensToAdd.add(screen);
    }

    public boolean isScreenVisible(@NotNull BaseScreen screen)
    {
        return screens.contains(screen, true);
    }

    public void startNewGame()
    {
        engine.removeAllEntities();
        try
        {
            SceneLoader.loadScene(engine, Gdx.files.internal("scenes/scene3.scene"));
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        catch (SceneLoader.LoadException e)
        {
            e.printStackTrace();
        }
    }

    /**
     * Exit the game. Cleans up all the resources
     */
    public void exit()
    {
        //TODO: Actually clean up the resources
        Gdx.app.exit();
    }

    public class InputMultiplexer implements InputProcessor
    {
        public boolean keyDown(int keycode)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().keyDown(keycode))
                {
                    return true;
                }
            }
            return false;
        }

        public boolean keyUp(int keycode)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().keyUp(keycode))
                {
                    return true;
                }
            }
            return false;
        }

        public boolean keyTyped(char character)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().keyTyped(character))
                {
                    return true;
                }
            }
            return false;
        }

        public boolean touchDown(int screenX, int screenY, int pointer, int button)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().touchDown(screenX, screenY, pointer, button))
                {
                    return true;
                }
            }
            return false;
        }

        public boolean touchUp(int screenX, int screenY, int pointer, int button)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().touchUp(screenX, screenY, pointer, button))
                {
                    return true;
                }
            }
            return false;
        }

        public boolean touchDragged(int screenX, int screenY, int pointer)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().touchDragged(screenX, screenY, pointer))
                {
                    return true;
                }
            }
            return false;
        }

        @Override
        public boolean mouseMoved(int screenX, int screenY)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().mouseMoved(screenX, screenY))
                {
                    return true;
                }
            }
            return false;
        }

        @Override
        public boolean scrolled(int amount)
        {
            for (int i = screens.size - 1; i >= 0; --i)
            {
                if (screens.get(i).getInputProcessor().scrolled(amount))
                {
                    return true;
                }
            }
            return false;
        }
    }
}

File: core/src/de/project/ice/IceGame.kt
package de.project.ice

import com.badlogic.gdx.*
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.I18NBundle
import de.project.ice.dialog.Dialog
import de.project.ice.dialog.Node
import de.project.ice.ecs.IceEngine
import de.project.ice.inventory.Inventory
import de.project.ice.screens.*
import de.project.ice.utils.Assets
import de.project.ice.utils.SceneLoader
import de.project.ice.utils.SceneWriter

import java.io.IOException
import java.util.*

open class IceGame : ApplicationAdapter() {
    private val screens = Array<BaseScreen>()
    private val screensToAdd = Array<BaseScreen>()
    private val screensToRemove = Array<Pair<BaseScreen, Boolean>>()
    protected val gameScreen: GameScreen by lazy { GameScreen(this, engine) }
    protected val cursorScreen: CursorScreen by lazy { CursorScreen(this) }
    protected val inputProcessor = InputMultiplexer()
    val engine: IceEngine by lazy { IceEngine(this) }
    val inventory = Inventory(this)
    val strings: I18NBundle by lazy { I18NBundle.createBundle(Gdx.files.internal("strings/ProjectIce")) }
    var isGamePaused = false
        private set
    var blockInteraction = false
        set(value) {
            field = value
            engine.controlSystem.setProcessing(!value)
        }
    var blockSaving = false
    var currentSaveSlot: Int = 0; private set

    override fun create() {
        I18NBundle.setSimpleFormatter(true)

        addScreen(gameScreen)
        addScreen(cursorScreen)

        init()

        Storage.SAVESTATE.load(currentSaveSlot)
    }

    protected open fun init() {
        addScreen(MainMenuScreen(this))
        Gdx.input.inputProcessor = inputProcessor

        SceneLoader.loadScene(engine, Gdx.files.internal("scenes/Hauptmenue.scene"))
    }

    fun pauseGame() {
        gameScreen.pauseGame()
        isGamePaused = true
    }

    fun resumeGame() {
        gameScreen.resumeGame()
        isGamePaused = false
    }

    fun save(slot: Int = currentSaveSlot): Boolean {
        var result = false
        if (!blockSaving) {
            val xmlState = SceneWriter.serializeToString(engine, engine.sceneProperties!!)
            Storage.SAVESTATE.put("__SAVESTATE__", xmlState)
            Storage.SAVESTATE.save(slot)
            result = true
        }
        currentSaveSlot = slot
        return result
    }

    fun load(slot: Int = currentSaveSlot, blendScreen: Boolean = false): Boolean {
        var result = false
        if (!blockSaving) {
            if (slot != currentSaveSlot)
                Storage.SAVESTATE.load(slot)

            if (hasSave()) {
                val xmlState = Storage.SAVESTATE.getString("__SAVESTATE__")
                if (blendScreen) {
                    engine.blendScreen(2.0f, Color.BLACK)

                    engine.timeout(2.0f) {
                        engine.entities.forEach { engine.removeEntity(it) }
                        engine.timeout(0f) {
                            try {
                                SceneLoader.loadScene(engine, xmlState)
                            } catch (e: IOException) {
                                e.printStackTrace()
                            } catch (e: SceneLoader.LoadException) {
                                e.printStackTrace()
                            }
                        }
                    }
                } else {
                    SceneLoader.loadScene(engine, xmlState)
                }
                result = true
            }
        }
        currentSaveSlot = slot
        return result
    }

    fun hasSave(slot: Int = currentSaveSlot): Boolean {
        if (!blockSaving)
            if (slot != currentSaveSlot)
                Storage.SAVESTATE.load(slot)

        return Storage.SAVESTATE.hasKey("__SAVESTATE__")
    }

    fun showDialog(node: Node, callback: ()->Unit = {}): DialogHandle {
        return DialogHandle(addScreen(DialogScreen(this, node).apply {
            this.callback = callback
        }), this)
    }

    fun showDialog(dialog: String, callback: ()->Unit = {}): DialogHandle {
        return showDialog(Dialog.load(Gdx.files.internal("dialog/$dialog.dlz")), callback)
    }


    fun showToastMessages(vararg messages: String) {
        val messageStrings = messages.map {
            try {
                strings.get(it)
            } catch (ex: MissingResourceException) {
                "$$$it$$"
            }
        }.toTypedArray()

        val messageScreen = screens.filter { it is MessageScreen }.firstOrNull() as MessageScreen?
        if (messageScreen != null)
            messageScreen.addMessages(*messageStrings)
        else
            addScreen(MessageScreen(this, *messageStrings))
    }


    fun showAndiMessages(vararg messages: String) {
        val start = Node()
        messages.fold(start) {
            node, msg -> Node().apply {
                type = Node.Type.Text
                color = Color.valueOf("#008fff")
                speaker = "Andi_Player"
                text = msg
                node.next = this
            }
        }
        showDialog(start)
    }

    override fun dispose() {
        for (screen in screens) {
            screen.hide()
            screen.dispose()
        }
        screens.clear()
    }

    override fun pause() {
        if (screens.size > 0) {
            screens.peek().pause()
        }
    }

    override fun resume() {
        if (screens.size > 0) {
            screens.peek().resume()
        }
    }

    var primaryCursor: CursorScreen.Cursor
        get() = cursorScreen.primaryCursor
        set(primaryCursor) {
            cursorScreen.primaryCursor = primaryCursor
        }

    var secondaryCursor: CursorScreen.Cursor
        get() = cursorScreen.secondaryCursor
        set(secondaryCursor) {
            cursorScreen.secondaryCursor = secondaryCursor
        }

    var cursorText: String
        get() = cursorScreen.cursorText
        set(cursorText) {
            cursorScreen.cursorText = cursorText
        }

    override fun render() {
        Assets.update()

        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT or GL20.GL_DEPTH_BUFFER_BIT)
        val delta = Gdx.graphics.deltaTime
        for (screen in Array(screens)) {
            screen.update(delta)
        }
        for (screen in screens) {
            screen.render()
        }
        for (screen in screensToAdd) {
            if (screens.contains(screen, true)) {
                continue
            }

            var index = -1
            for (i in 0..screens.size - 1) {
                if (screens.get(i).priority < screen.priority) {
                    index = i
                    break
                }
            }
            if (index == -1) {
                if (screens.size > 0) {
                    screens.peek().pause()
                }
                screens.add(screen)
                screen.show()
            } else {
                screens.insert(index, screen)
            }
            screen.resize(Gdx.graphics.width, Gdx.graphics.height)
        }
        screensToAdd.clear()
        for (pair in screensToRemove.filter { screens.contains(it.first, true) }) {
            if (pair === screens.peek()) {
                screens.pop()
                screens.peek().resume()
            } else {
                screens.removeValue(pair.first, true)
            }
            pair.first.hide()
            if (pair.second) {
                pair.first.dispose()
            }
        }
        screensToRemove.clear()
    }

    override fun resize(width: Int, height: Int) {
        for (screen in screens) {
            screen.resize(width, height)
        }
    }

    @JvmOverloads fun removeScreen(screen: BaseScreen, dispose: Boolean = true) {
        screensToRemove.add(Pair(screen, dispose))
    }

    fun <T: BaseScreen> addScreen(screen: T): T{
        screensToAdd.add(screen)
        return screen
    }

    fun isScreenVisible(screen: BaseScreen): Boolean {
        return screens.contains(screen, true)
    }

    fun startNewGame() {
        Storage.SAVESTATE.clear()
        screens.filter { it is MainMenuScreen }.forEach { removeScreen(it) }
        engine.blendScreen(2.0f, Color.BLACK)

        engine.timeout(2.0f) {
            engine.entities.forEach { engine.removeEntity(it) }
            engine.timeout(0f) {
                try {
                    SceneLoader.loadScene(engine, Gdx.files.internal("scenes/scene1.scene"))
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: SceneLoader.LoadException) {
                    e.printStackTrace()
                }
            }
        }
    }

    /**
     * Exit the game. Cleans up all the resources
     */
    fun exit() {
        //TODO: Actually clean up the resources
        Gdx.app.exit()
    }

    inner class InputMultiplexer : InputProcessor {
        override fun keyDown(keycode: Int): Boolean {
            if (keycode == Input.Keys.ENTER && Gdx.input.isKeyPressed(Input.Keys.ALT_LEFT)) {
                if (Gdx.graphics.isFullscreen)
                    Gdx.graphics.setWindowedMode(1920, 1080)
                else for (mode in Gdx.graphics.getDisplayModes(Gdx.graphics.primaryMonitor)) {
                        if (mode.width == 1920 && mode.height == 1080) {
                            Gdx.graphics.setFullscreenMode(mode)
                            break
                        }
                    }
                return true
            }
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.keyDown(keycode)) {
                    return true
                }
            }
            return false
        }

        override fun keyUp(keycode: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.keyUp(keycode)) {
                    return true
                }
            }
            return false
        }

        override fun keyTyped(character: Char): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.keyTyped(character)) {
                    return true
                }
            }
            return false
        }

        override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.touchDown(screenX, screenY, pointer, button)) {
                    return true
                }
            }
            return false
        }

        override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.touchUp(screenX, screenY, pointer, button)) {
                    return true
                }
            }
            return false
        }

        override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.touchDragged(screenX, screenY, pointer)) {
                    return true
                }
            }
            return false
        }

        override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.mouseMoved(screenX, screenY)) {
                    return true
                }
            }
            return false
        }

        override fun scrolled(amount: Int): Boolean {
            for (i in screens.size - 1 downTo 0) {
                if (screens.get(i).inputProcessor.scrolled(amount)) {
                    return true
                }
            }
            return false
        }
    }
}

class DialogHandle internal constructor(private val screen: DialogScreen, private val game: IceGame) {
    fun cancel() {
        game.removeScreen(screen)
    }
}
